src/Temporalio.Extensions.Hosting/ServiceProviderAccessor.cs

```diff
@@ -0,0 +1,42 @@
+using System;
+using System.Threading;
+
+namespace Temporalio.Extensions.Hosting
+{
+    /// <summary>
+    /// Provides an implementation of <see cref="IServiceProvider" /> based on
+    /// the current execution context.
+    /// </summary>
+    public class ServiceProviderAccessor : IServiceProviderAccessor
+    {
+        private static readonly AsyncLocal<ServiceProviderHolder> ServiceProviderCurrent = new();
```

Comment:
A `System.Collections.Concurrent.ConcurrentDictionary` that is keyed on the Activity's unique ID could be a simpler replacement for this, I think, but I'm not sure when `ActivityExecutionContext.Current` is set and I think that would be implicitly needed by `ServiceProvider.set`. I would expect it would definitely be available when `ServiceProvider.get` is expected to work.